price-estimator/src/orderbook.rs

```diff
@@ -1,35 +1,83 @@
-use core::orderbook::StableXOrderBookReading;
-use pricegraph::Pricegraph;
-use std::time::{Duration, SystemTime};
+use crate::solver_rounding_buffer;
+use core::{
+    models::TokenId, orderbook::StableXOrderBookReading,
+    price_estimation::price_source::PriceSource,
+};
+use pricegraph::{Pricegraph, TokenPair};
+use std::{
+    sync::Arc,
+    time::{Duration, SystemTime},
+};
 use tokio::sync::RwLock;
 
+pub type DynPriceSource = Box<dyn PriceSource + Send + Sync>;
+
 /// Access and update the pricegraph orderbook.
-pub struct Orderbook<T> {
-    orderbook_reading: T,
+pub struct Orderbook {
+    orderbook_reading: Arc<dyn StableXOrderBookReading>,
+    price_source: DynPriceSource,
+    all_tokens: Vec<TokenId>,
     pricegraph: RwLock<Pricegraph>,
+    rounding_buffer_factor: f64,
 }
 
-impl<T> Orderbook<T> {
-    pub fn new(orderbook_reading: T) -> Self {
+impl Orderbook {
+    pub fn new(
+        orderbook_reading: Arc<dyn StableXOrderBookReading>,
+        price_source: DynPriceSource,
+        all_tokens: Vec<TokenId>,
+        rounding_buffer_factor: f64,
+    ) -> Self {
+        assert!(rounding_buffer_factor.is_finite());
+        assert!(rounding_buffer_factor >= 0.0);
         Self {
             orderbook_reading,
+            price_source,
+            all_tokens,
             pricegraph: RwLock::new(Pricegraph::new(std::iter::empty())),
+            rounding_buffer_factor,
         }
     }
 
-    pub async fn get_pricegraph(&self) -> Pricegraph {
+    pub async fn pricegraph(&self) -> Pricegraph {
         self.pricegraph.read().await.clone()
     }
-}
 
-impl<T: StableXOrderBookReading> Orderbook<T> {
+    /// See solver_rounding_buffer::rounding_buffer. This is used to adjust amounts in api queries
+    /// to adapt to how the solver sees them.
+    /// If any price is not available `1` is used as a fallback.
+    pub async fn rounding_buffer(&self, token_pair: TokenPair) -> f64 {
+        let fee_token = TokenId(0);
+        let sell_token = TokenId(token_pair.sell);
+        let buy_token = TokenId(token_pair.buy);
+        let prices = self
+            .price_source
+            .get_prices(&[fee_token, sell_token, buy_token])
+            .await
+            .unwrap_or_default();
+        let get_price = |token_id| prices.get(&token_id).copied().unwrap_or(1) as f64;
```

Comment:
Both here and `solver_rounding_buffer` we need to choose some value for the price if we don't have it. It is unclear what reasonable is. Now that you mention it more reasonable would probably be to reuse the fee price and if we also don't have the fee price then use 1 for that.

Reply:
Shouldn't this consider the decimals of the token? What would be a pessimistic default here? A high price or a low price? We could also just fall back to 0.1% rounding in case there is no price available.